File: scripts/ocr_image.m
#import <Foundation/Foundation.h>
#import <Vision/Vision.h>
#import <CoreGraphics/CoreGraphics.h>
#import <ImageIO/ImageIO.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        if (argc < 2) {
            fprintf(stderr, "usage: ocr_image <image>\n");
            return 2;
        }
        NSString *path = [NSString stringWithUTF8String:argv[1]];
        NSURL *url = [NSURL fileURLWithPath:path];
        CGImageSourceRef source = CGImageSourceCreateWithURL((__bridge CFURLRef)url, NULL);
        if (!source) {
            fprintf(stderr, "failed to load image source\n");
            return 1;
        }
        CGImageRef image = CGImageSourceCreateImageAtIndex(source, 0, NULL);
        CFRelease(source);
        if (!image) {
            fprintf(stderr, "failed to load image\n");
            return 1;
        }
        VNRecognizeTextRequest *request = [[VNRecognizeTextRequest alloc] initWithCompletionHandler:nil];
        request.recognitionLevel = VNRequestTextRecognitionLevelAccurate;
        request.usesLanguageCorrection = YES;
        request.recognitionLanguages = @[@"zh-Hans", @"en-US"];
        request.minimumTextHeight = 0.008;
        VNImageRequestHandler *handler = [[VNImageRequestHandler alloc] initWithCGImage:image options:@{}];
        NSError *error = nil;
        BOOL ok = [handler performRequests:@[request] error:&error];
        CGImageRelease(image);
        if (!ok) {
            if (error) {
                fprintf(stderr, "error domain=%s code=%ld desc=%s\n", [[error domain] UTF8String], (long)[error code], [[error localizedDescription] UTF8String]);
            } else {
                fprintf(stderr, "performRequests failed with nil error\n");
            }
            return 1;
        }
        NSArray<VNRecognizedTextObservation *> *results = request.results ?: @[];
        NSArray *sorted = [results sortedArrayUsingComparator:^NSComparisonResult(VNRecognizedTextObservation *a, VNRecognizedTextObservation *b) {
            CGFloat dy = a.boundingBox.origin.y + a.boundingBox.size.height / 2.0 - (b.boundingBox.origin.y + b.boundingBox.size.height / 2.0);
            if (fabs(dy) > 0.01) {
                return dy > 0 ? NSOrderedAscending : NSOrderedDescending;
            }
            if (a.boundingBox.origin.x < b.boundingBox.origin.x) return NSOrderedAscending;
            if (a.boundingBox.origin.x > b.boundingBox.origin.x) return NSOrderedDescending;
            return NSOrderedSame;
        }];
        fprintf(stderr, "results=%lu\n", (unsigned long)[sorted count]);
        for (VNRecognizedTextObservation *obs in sorted) {
            VNRecognizedText *txt = [[obs topCandidates:1] firstObject];
            if (txt && txt.string) {
                printf("%s\n", [txt.string UTF8String]);
            }
        }
    }
    return 0;
}

File: scripts/ocr_image.swift
import Foundation
import Vision
import CoreGraphics
import ImageIO

if CommandLine.arguments.count < 2 {
    fputs("usage: swift ocr_image.swift <image>\n", stderr)
    exit(2)
}

let path = CommandLine.arguments[1]
let url = URL(fileURLWithPath: path)
guard let source = CGImageSourceCreateWithURL(url as CFURL, nil),
      let image = CGImageSourceCreateImageAtIndex(source, 0, nil) else {
    fputs("failed to load image\n", stderr)
    exit(1)
}

let request = VNRecognizeTextRequest()
request.recognitionLevel = .accurate
request.usesLanguageCorrection = true
request.recognitionLanguages = ["zh-Hans", "en-US"]
request.minimumTextHeight = 0.008

let handler = VNImageRequestHandler(cgImage: image, options: [:])
try handler.perform([request])

let lines = (request.results ?? []).compactMap { observation -> (CGFloat, CGFloat, String)? in
    guard let candidate = observation.topCandidates(1).first else { return nil }
    return (observation.boundingBox.midY, observation.boundingBox.minX, candidate.string)
}
.sorted { a, b in
    if abs(a.0 - b.0) > 0.01 {
        return a.0 > b.0
    }
    return a.1 < b.1
}
.map { $0.2 }

print(lines.joined(separator: "\n"))
